### PR TITLE
Use of CRTM UFO

### LINKED ISSUE #246: Create an atmospheric Geometry in obs space
**Create an optional Atmospheric Geometry in Obs Space**
- Get "grid" size, lat, lon from obs file
- Pass atm grid spec to the soca::Geometry
- Adjust Geometry methods as necessary

**Tests**
- [ ] Backward compatible: Create ocean ice only geometry (no atmosphere)
- [ ] Exercise the new feature and create the atm geometry in addition to the ocean and ice

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description
**Oil spill is contained!** None of the atmospheric variables are mentioned anywhere in the code (except yaml files), so far ...
Revisiting aborted [PR#255](https://github.com/JCSDA/soca/pull/255.) ... and more. This feature branch adds the possibility of using the CRTM ufo in hofx3d (not tested yet with hofx):
- Atmospheric geovals are pre-computed (fv3-jedi) and read inside of getvalue
- Ocean surface (just sst for now) is provided by soca through getvalue.

Is a change of answers expected from this PR? **NO but overflow issue in adt hofx. No clue why. @travissluka and @aerorahul said they'll fix it.**

### Issue(s) addressed
- closes #246 
- Partially  addresses #248, #247, #254 

### What's not too flash
The atmospheric "geometry" and "state" are part of the soca geometry. Not pretty but I can't think of a better way that would work.

## Testing
How were these changes tested? **hofx3dcrtm and geometryatm ctest**
What compilers / HPCs was it tested with? **gcc only**
Are the changes covered by ctests? (If not, tests should be added first.) **YES geometryatm test**

## Dependencies
Need the CRTM

### CODE REVIEW

**Comment by travissluka** [comment]
tests are failing because CRTM is not added to travisCI (so, ha, both Travises dont want CRTM).

Let me do a quick PR first that is needed for letting travisCI config set the bundle-wide ecbuild options .... while we continue to debate the merits of requiring CRTM.

I vote for adding in if statements to the cmakelist so that the crtm tests are skipped if crtm is not present (see the way ufo ctests are setup). We'll have crtm on by default on travisci... and off by default on my and @aerorahul 's laptops. everyone wins

**Comment by travissluka** [comment]
@guillaumevernieres yeah, that hofx is expected to fail on intel. I don't know what i was thinking enabling SIGFPE trap on that given that the model runs for hofx.

**Review by travissluka — CHANGES_REQUESTED** [review]
after JCSDA/soca/pull/260 gets merged, modify the CMakeLists.txt to wrap the ctests that need CRTM with `if( ${CRTM_FOUND} )`

**Comment by guillaumevernieres** [comment]
@travissluka I must have screwed up somewhere, the crtm test didn't run. I'll look at it later tonight or tomorrow.

**Comment by travissluka** [comment]
I think the problem is that soca isn't seeing that crtm is available because it isn't listed as a dependency in the main `CMakeLists.txt`. This doesn't matter if building as a normal bundle since ufo pulls in crtm, but the TravisCI tests build each repo separately.

try adding the line
`ecbuild_use_package( PROJECT crtm VERSION 2.2.3)`

Also, there's probably no harm in leaving the `geometryatm` test enabled even if crtm is not present

**Comment by guillaumevernieres** [comment]
@travissluka OK, trying again.

File: CMakeLists.txt [hunk 1]
```diff
@@ -66,6 +66,10 @@ include_directories( ${IODA_INCLUDE_DIRS} )
 ecbuild_use_package( PROJECT ufo VERSION 0.1.0 REQUIRED )
 include_directories( ${UFO_INCLUDE_DIRS} )
 
+# ufo
+ecbuild_use_package( PROJECT ufo VERSION 0.1.0 REQUIRED )
+include_directories( ${UFO_INCLUDE_DIRS} )
+
```
Comment (travissluka) [inline CMakeLists.txt]: oooh, double UFO !

**Comment by travissluka** [comment]
@guillaumevernieres can't find the binary file now... because we aren't building everything as a single bundle. I had the same issue with files from oops, I'll lookup the syntax needed and get back to you...

**Comment by guillaumevernieres** [comment]
Still no cigar, but closer. Looks like the .bin for the crtm were not found ...  Ha, just seeing your post above.  Nice, thanks for your help.

File: CMakeLists.txt (hunk 1 above)
Comment (guillaumevernieres) [reply CMakeLists.txt]: dang, I knew the copy paste went badly ... Will wait until you send me a solution for the lfs files before pushing the fix.

**Comment by guillaumevernieres** [comment]
Thanks @travissluka

File: CMakeLists.txt [hunk 2]
```diff
@@ -79,6 +79,10 @@ ecbuild_find_mpi( COMPONENTS CXX Fortran REQUIRED )
 ecbuild_include_mpi()
 link_libraries(${MPI_CXX_LIBRARIES} ${MPI_Fortran_LIBRARIES})
 
+# crtm
+ecbuild_use_package( PROJECT crtm VERSION 2.2.3 )
+include_directories( ${CRTM_INCLUDE_DIRS} )
```
Comment (travissluka) [inline CMakeLists.txt]:
```suggestion
```

File: src/soca/State/State.cc [hunk 3]
```diff
@@ -84,7 +84,17 @@ namespace soca {
   void State::getValues(const ufo::Locations & locs,
                         const oops::Variables & vars,
                         ufo::GeoVaLs & cols) const {
+    if (fields_->geometry()->getAtmInit())
+      {
+        // Get atm geovals
+        // The variables in vars that are also defined in soca will be
+        // over-written in the interpolation call bellow
+        getValuesFromFile(locs, vars, cols);
+        Log::trace() << cols << std::endl;
```
Comment (travissluka) [inline src/soca/State/State.cc]: remove, or make the output informative

File: src/soca/State/State.cc [hunk 4]
```diff
@@ -84,7 +84,17 @@ namespace soca {
...
+        }
+    // Get ocean geovals
     fields_->getValues(locs, vars, cols);
+    Log::trace() << cols << std::endl;
```
Comment (travissluka) [inline src/soca/State/State.cc]: same as above

**Review by travissluka — CHANGES_REQUESTED** [review]

File: src/soca/Fields/soca_interpfields_mod.F90 [hunk 5]
```diff
@@ -188,6 +190,11 @@ subroutine interp(fld, locs, vars, geoval, horiz_interp, traj)
      case ("sea_surface_temperature")
         fld3d(isc:iec,jsc:jec,1) = fld%tocn(isc:iec,jsc:jec,1)
 
+     ! TODO: Move unit change elsewhere, check if surface_temperature_where_sea
```
Comment (travissluka) [inline src/soca/Fields/soca_interpfields_mod.F90:193]: might want to create a github issue to remind us.... we had discussed before that all of our ocean temp variables should be in K to comply with CF conventions. But of course that requires changing ufo at the same time... a good project for when someone is bored and doesn't want to do real work.

**Review by travissluka — APPROVED** [review]
Less of a Frankenstein than I was expecting 👍 

I tested on intel19 as well and all looks good.

**Pull request merged**